src/compiler/binder.ts

```diff
@@ -567,174 +570,501 @@ namespace ts {
+            return <FlowAssignment>{
+                kind: FlowKind.Assignment,
+                antecedent,
+                node
+            };
+        }
 
-            // bind expression (don't affect reachability)
-            bind(n.expression);
+        function finishFlow(flow: FlowNode): FlowNode {
+            while (flow.kind === FlowKind.Label) {
+                const antecedents = (<FlowLabel>flow).antecedents;
+                if (!antecedents) {
+                    return unreachableFlow;
+                }
+                if (antecedents.length > 1) {
+                    break;
+                }
+                flow = antecedents[0];
+            }
+            return flow;
+        }
 
-            bind(n.thenStatement);
-            if (n.elseStatement) {
-                const preElseState = currentReachabilityState;
-                currentReachabilityState = ifFalseState;
-                bind(n.elseStatement);
-                currentReachabilityState = or(currentReachabilityState, preElseState);
+        function isStatementCondition(node: Node) {
+            const parent = node.parent;
+            switch (parent.kind) {
+                case SyntaxKind.IfStatement:
+                case SyntaxKind.WhileStatement:
+                case SyntaxKind.DoStatement:
+                    return (<IfStatement | WhileStatement | DoStatement>parent).expression === node;
+                case SyntaxKind.ForStatement:
+                case SyntaxKind.ConditionalExpression:
+                    return (<ForStatement | ConditionalExpression>parent).condition === node;
             }
-            else {
-                currentReachabilityState = or(currentReachabilityState, ifFalseState);
+            return false;
+        }
+
+        function isLogicalExpression(node: Node) {
+            while (true) {
+                if (node.kind === SyntaxKind.ParenthesizedExpression) {
+                    node = (<ParenthesizedExpression>node).expression;
+                }
+                else if (node.kind === SyntaxKind.PrefixUnaryExpression && (<PrefixUnaryExpression>node).operator === SyntaxKind.ExclamationToken) {
+                    node = (<PrefixUnaryExpression>node).operand;
+                }
+                else {
+                    return node.kind === SyntaxKind.BinaryExpression && (
+                        (<BinaryExpression>node).operatorToken.kind === SyntaxKind.AmpersandAmpersandToken ||
+                        (<BinaryExpression>node).operatorToken.kind === SyntaxKind.BarBarToken);
+                }
             }
         }
 
-        function bindReturnOrThrow(n: ReturnStatement | ThrowStatement): void {
-            // bind expression (don't affect reachability)
-            bind(n.expression);
-            if (n.kind === SyntaxKind.ReturnStatement) {
+        function isTopLevelLogicalExpression(node: Node): boolean {

```

Comment:
is it just for logical expressions or for other expressions too? cause I was wondering (language agnostically) what happens if something always raises an exception but isn't a throw statement (like dividing by zero), should one be clever enough to handle it? And should they also resolve constant values first or other statements that are calculatable at checking time, then check for such a case? Or would it be too far fetched, behaving like an AI coprogrammer?